druid/src/widget/focus.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 The druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A focus widget.
+
+use druid::widget::prelude::*;
+
+use druid::{commands, Data, FocusNode, HotKey, KbKey, Point, Rect, SysMods, Widget, WidgetPod};
+
+#[derive(Debug, Clone)]
+pub(crate) struct FocusController {
+    focus_node: FocusNode,
+    previous_focus_node: FocusNode,
+    requested_focus: bool,
+    pub(crate) auto_focus: bool,
+}
+
+impl FocusController {
+    pub(crate) fn new() -> Self {
+        Self {
+            focus_node: FocusNode::empty(),
+            previous_focus_node: FocusNode::empty(),
+            requested_focus: false,
+            auto_focus: false,
+        }
+    }
+
+    pub(crate) fn event_start(&mut self, ctx: &mut EventCtx) {
+        self.previous_focus_node = ctx.focus_node();
+        ctx.set_focus_node(self.focus_node);
+    }
+
+    pub(crate) fn event_end(&mut self, ctx: &mut EventCtx, event: &Event) {
+        match event {
+            Event::MouseDown(_) => {
+                ctx.request_focus();
+                ctx.request_paint();
+            }
+            Event::KeyDown(key_event) if !ctx.is_handled => {
+                match key_event {
+                    // Tab and shift+tab
+                    k_e if HotKey::new(None, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_next();
+                    }
+                    k_e if HotKey::new(SysMods::Shift, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_prev();
```

Comment:
we'll want some way to make this behaviour configurable at some point, I think? but no worries for now.

Reply:
I agree, I think it would be great if we could configure it.